### ISSUE TITLE
The questions about the gradient between the leader agent and the follower agent

### ISSUE THREAD

**Issue by ZoraTian** [laonahongchen/Bilevel-Optimization-in-Coordination-Game#2]
Hey, I have read your paper and I thought it is very useful for me. But I have some questions about the gradients. I find the linkage between the leader agent and the follower agent is all numpy format, and I don't understand the propagation of the gradient such as Q1 and π2 in eq. (12) in the paper.
Could you answer my question？

**Comment by laonahongchen** [comment]
Hi @ZoraTian , thank you for your interest in our paper. To help us understand your question better, can you be more specific about which part you think should propagate the gradient?

Specifically, if I remember correctly, Q1 as a Q function, purely takes state and joint actions as the input and provides a Q-value estimation, and should not need any gradient with respect to the action **distribution** $\pi_2$, while $\pi_2$ is already taking the action of the leader, again does not require any gradient from Q1. And Eq. 12 is about how the action is selected, not regarding how the update is done. The update rule for Q1 is Eq. 10, which is similar to other Q-function-based MARL algorithms that take joint actions like Qmix, and naturally handled by training on trajectories from replay buffers. 

Feel free to let me know if you want to follow up by providing more specific questions. While it is acknowledged long since the paper was published, and thus it might not be easy even for us to go back to the exact code implementations, (especially given that this repo is based on tensorflow which is much less popular right now), I am personally happy to help with clarifying how the algorithm works.

**Comment by laonahongchen** [comment]
We are closing this issue due to inactivity. You are welcome to reopen the issue if you have follow-up questions.
